<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro" name="rover">
    <link name="base_link">
        <visual>
            <geometry>
                <box size="1.0 0.7 0.1" />
            </geometry>
            <material name="red">
                <color rgba="1 0 0 1"/>
            </material>
        </visual>
        <collision>
            <geometry>
                <box size="1.0 1.0 1.0" />
            </geometry>
        </collision>
        <inertial>
            <mass value="0.0"/>
            <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0"/>
            <inertia ixx="0.0" ixy="0.0" ixz="0.0" iyy="0.0" iyz="0.0" izz="0.0"/>
        </inertial>
    </link>
    <link name="base_footprint">
    </link>
    <joint name="base_joint" type="fixed">
        <parent link="base_footprint"/>
        <child link="base_link"/>
        <origin xyz="0 0 0.2" rpy="0 0 0"/>
        <axis xyz="0 0 0.2"/>
    </joint>


    <xacro:property name="pi" value="3.14"/>

    <xacro:macro name="wheel" params="name x y">
        <link name="wheel_${name}_link">
            <visual>
                <geometry>
                    <cylinder radius="0.15" length="0.1"/>
                </geometry>
                <material name="LightGrey">
                    <color rgba="0.7 0.7 0.7 1.0"/>
                </material>
            </visual>
            <collision>
                <geometry>
                    <cylinder radius="1.0" length="1.0"/>
                </geometry>
            </collision>
            <inertial>
                <mass value="0.0"/>
                <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0"/>
                <inertia ixx="0.0" ixy="0.0" ixz="0.0" iyy="0.0" iyz="0.0" izz="0.0"/>
            </inertial>
        </link>

        <joint name="wheel_${name}_joint" type="continuous">
            <parent link="base_link"/>
            <child link="wheel_${name}_link"/>
            <origin xyz="${x} ${y} 0.0" rpy="${pi/2} 0 0"/>
            <axis xyz="0 0 1.0"/>
            <limit lower="0.0" upper="0.0" effort="0.1" velocity="0.1"/>
        </joint>

    </xacro:macro>
    
    <xacro:wheel name="front_right" x="-0.25" y="0.4"/>
    <xacro:wheel name="front_left" x="-0.25" y="-0.4"/>
    <xacro:wheel name="rear_right" x="0.25" y="0.4"/>
    <xacro:wheel name="rear_left" x="0.25" y="-0.4"/>

</robot>